component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -2704,4 +2807,228 @@ private boolean isOTPMismatched(AuthenticationContext context) {
     }
+
+    /**
+     * Check whether the Identifier first page is initialized from within the email OTP authenticator.
+     *
+     * @param context Authentication context
+     * @return whether the Identifier first page is initialized from within the email OTP authenticator
+     */
+    private boolean isIDFInitiatedFromEmailOTP(AuthenticationContext context) {
+
+        return Boolean.TRUE.equals(context.getProperty(
+                EmailOTPAuthenticatorConstants.IS_IDF_INITIATED_FROM_AUTHENTICATOR));
+    }
+
+    /**
+     * This method is used to redirect the user to the username entering page (IDF: Identifier first).
+     *
+     * @param context  The authentication context
+     * @param response Response
+     * @throws AuthenticationFailedException AuthenticationFailedException
+     */
+    private void redirectUserToIDF(HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        String loginPage = ConfigurationFacade.getInstance().getAuthenticationEndpointURL();
+        String queryParams = context.getContextIdIncludedQueryParams();
+        try {
+            if (log.isDebugEnabled()) {
+                String logMsg = String.format("Redirecting to identifier first flow since "
+                        + "last authenticated user is null in SP: %s", context.getServiceProviderName());
+                log.debug(logMsg);
+            }
+            response.sendRedirect(loginPage + ("?" + queryParams) + "&"
+                    + EmailOTPAuthenticatorConstants.AUTHENTICATORS
+                    + EmailOTPAuthenticatorConstants.IDF_HANDLER_NAME + ":"
+                    + EmailOTPAuthenticatorConstants.LOCAL_AUTHENTICATOR);
+        } catch (IOException e) {
+            throw new AuthenticationFailedException("Error while redirecting to the login page.", e);
+        }
+    }
+
+    /**
+     * Check whether the username is obtained in the previous step.
+     *
+     * @param context AuthenticationContext.
+     * @return boolean value.
+     */
+    private boolean isSubjectAttributeObtained(AuthenticationContext context) {
+        if (context.getLastAuthenticatedUser() != null
+                && context.getLastAuthenticatedUser().getUserName() != null) {
+            return true;
+        }
+        Map<Integer, StepConfig> stepConfigMap = context.getSequenceConfig().getStepMap();
+        AuthenticatedUser authenticatedUser;
+        String username;
+        for (StepConfig stepConfig : stepConfigMap.values()) {
+            authenticatedUser = stepConfig.getAuthenticatedUser();
+            if (authenticatedUser != null && stepConfig.isSubjectAttributeStep()) {
+                username = authenticatedUser.toFullQualifiedUsername();
+                if (username != null) {
+                    return true;
+                }
+            }
+        }
+        return false;
```

Comment:
let's try to use stream APIs :)
```suggestion
        Optional<Boolean> isValidUser = stepConfigMap.values().stream()
        .filter(stepConfig -> stepConfig.getAuthenticatedUser() != null && stepConfig.isSubjectAttributeStep())
        .map(stepConfig -> stepConfig.getAuthenticatedUser().toFullQualifiedUsername())
        .filter(Objects::nonNull)
        .findFirst()
        .isPresent();

return isValidUser.orElse(false);
```

Reply:
Removed unused method.